File: fenwicks/3187-peaks-in-array.go
package fenwicks

// https://leetcode.com/problems/peaks-in-array/

func countOfPeaks(nums []int, queries [][]int) []int {
	n := len(nums)
	bitree := make([]int, n+1)
	isPeak := func(i int) bool {
		return i > 0 && i < n-1 && nums[i-1] < nums[i] && nums[i] > nums[i+1]
	}
	update := func(index int, value int) {
		index++
		for ; index <= n; index += index & -index {
			bitree[index] += value
		}
	}
	query := func(index int) int {
		var total int

		index++
		for ; index > 0; index -= index & -index {
			total += bitree[index]
		}

		return total
	}

	for i := range n {
		if isPeak(i) {
			update(i, 1)
		}
	}

	var (
		counts                    []int
		index, value, left, right int
		curr                      bool
	)

	for i := range queries {
		if queries[i][0] == 2 {
			index, value = queries[i][1], queries[i][2]
			before := []bool{isPeak(index - 1), isPeak(index), isPeak(index + 1)}

			nums[index] = value
			for i := range before {
				curr = isPeak(index + i - 1)
				if before[i] && !curr {
					update(index+i-1, -1)
				} else if !before[i] && curr {
					update(index+i-1, 1)
				}
			}
		} else {
			left, right = queries[i][1], queries[i][2]
			if left < right {
				// [left+1, right-1]
				counts = append(counts, query(right-1)-query(left))
			} else {
				counts = append(counts, 0)
			}
		}
	}

	return counts
}

File: fenwicks/3187-peaks-in-array.py
from typing import List
import unittest

# https://leetcode.com/problems/peaks-in-array/

# python3 -m unittest fenwicks/3187-peaks-in-array.py


class Solution(unittest.TestCase):
    def countOfPeaks(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        n = len(nums)
        bitree = [0] * (n + 1)

        def update(index: int, value: int):
            index += 1
            while index <= n:
                bitree[index] += value
                index += index & -index

        def query(index: int) -> int:
            total = 0
            index += 1
            while index > 0:
                total += bitree[index]
                index -= index & -index
            return total

        def is_peak(i: int) -> bool:
            return 0 < i < n - 1 and nums[i - 1] < nums[i] > nums[i + 1]

        for i in range(n):
            if is_peak(i):
                update(i, 1)

        counts: List[int] = []
        for t, first, second in queries:
            if t == 1:
                left, right = first, second
                counts.append(query(right - 1) - query(left) if left < right else 0)
            else:
                index, value = first, second
                before = [is_peak(index - 1), is_peak(index), is_peak(index + 1)]
                nums[index] = value
                for i in range(len(before)):
                    curr = is_peak(index + i - 1)
                    if before[i] and not curr:
                        update(index + i - 1, -1)
                    elif not before[i] and curr:
                        update(index + i - 1, 1)

        return counts

    def test(self):
        for nums, queries, expected in [
            ([3, 1, 4, 2, 5], [[2, 3, 4], [1, 0, 4]], [0]),
            ([4, 1, 4, 2, 1, 5], [[2, 2, 4], [1, 0, 2], [1, 0, 4]], [0, 1]),
            ([4, 9, 4, 10, 7], [[2, 3, 2], [2, 1, 3], [1, 2, 3]], [0]),
            ([5, 4, 8, 6], [[1, 2, 2], [1, 1, 2], [2, 1, 6]], [0, 0]),
        ]:
            output = self.countOfPeaks(nums, queries)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
